plugins/outputs/cloudwatchlogs/internal/pusher/sender.go

```diff
@@ -110,27 +110,28 @@ func (s *sender) Send(batch *logEventBatch) {
 			s.logger.Errorf("Aws error received when sending logs to %v/%v: %v", batch.Group, batch.Stream, awsErr)
 		}
 
-		// retry wait strategy depends on the type of error returned
-		var wait time.Duration
-		if chooseRetryWaitStrategy(err) == retryLong {
-			wait = retryWaitLong(retryCountLong)
-			retryCountLong++
-		} else {
-			wait = retryWaitShort(retryCountShort)
-			retryCountShort++
-		}
+		// Update retry metadata in the batch
+		batch.updateRetryMetadata(err)
 
-		if time.Since(startTime)+wait > s.RetryDuration() {
-			s.logger.Errorf("All %v retries to %v/%v failed for PutLogEvents, request dropped.", retryCountShort+retryCountLong-1, batch.Group, batch.Stream)
+		// Check if retry would exceed max duration
+		totalRetries := batch.retryCountShort + batch.retryCountLong - 1
+		if batch.isExpired(s.RetryDuration()) || batch.nextRetryTime.After(batch.startTime.Add(s.RetryDuration())) {
```

Comment:
Wont the 2nd condition in this line always be triggered before `isExpired` returns true?

Also, can the second part be moved into `isExpired()`? Would the following achieve the same?
```
maxWaitTime := b.startTime.Add(maxRetryDuration)
b.nextRetryTime.After(maxWaitTime) || time.Now().After(maxWaitTime)
```

If this is equivalent to your logic, then even here we can see that the first part will always be true before the second part can be true. Since nextRetryTime in n-1 iteration will be more than time.Now() in the nth iteration.

Reply:
I guess the time.Now() check in isExpired() is useful for the subsequent PRs when the RetryHeapProcessor comes into play?

In that case, should the condition here just be:
```if batch.nextRetryTime.After(batch.startTime.Add(s.RetryDuration()))```